### PR TITLE
Bug 1484971 Refactor Amplitude-specific logic out of HttpSink

### PR DESCRIPTION
**Pull request by jklukas**
There's also a preliminary commit https://github.com/mozilla/telemetry-streaming/pull/174/commits/2b88c05995c7efd6bb19f438da266c996e0e7289 that makes naming of class and package consistent for HttpSink.

The general strategy for the main commit is to refactor HttpSink to be an abstract base class, and then specific APIs correspond to case classes extending HttpSink.

A final commit refactors `BatchHttpSink` to also extend `HttpSink`.

### CODE REVIEW

File: src/main/scala/com/mozilla/telemetry/sinks/CrashesBatchHttpSink.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package com.mozilla.telemetry.sinks
+
+import scalaj.http.HttpRequest
+
+import scala.collection.mutable.ArrayBuffer
+
+case class CrashesBatchHttpSink(
```
Comment (jklukas) [inline src/main/scala/com/mozilla/telemetry/sinks/CrashesBatchHttpSink.scala]: Used to be BatchHttpSink. It's renamed here and refactored to extend HttpSink (used to be standalone).

File: src/main/scala/com/mozilla/telemetry/sinks/HttpSink.scala [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+object HttpSink {
+  val TimeoutPseudoCode: Int = -1
+  val ErrorPseudoCode: Int = -2
+  val OK = 200
+  val TooManyRequests = 429
+  val InternalServerError = 500
+  val BadGateway = 502
+  val ServiceUnavailable = 503
+  val GatewayTimeout = 504
+  val RetryCodes: Set[Int] = Set(TimeoutPseudoCode, TooManyRequests, InternalServerError, BadGateway, ServiceUnavailable, GatewayTimeout)
```
[the remark this answers is not in the document]
Comment (jklukas) [reply src/main/scala/com/mozilla/telemetry/sinks/HttpSink.scala]: Yes, these codes are called out specifically in Amplitude's docs, but they're not Amplitude-specific. 429 should always indicate that you're being throttled and should be handled by backing off and retrying, for example.

From my understanding, this list of codes is the complete set you'd want to retry on for any API.

I will add some more comments to that effect, though.

File: src/main/scala/com/mozilla/telemetry/sinks/HttpSink.scala [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+/**
+  * An abstract base class for sending data to an HTTP API.
+  *
+  * @tparam T type of the values passed in to httpSendMethod
+  */
+abstract class HttpSink[T]() extends ForeachWriter[T] {
```
[the remark this answers is not in the document]
Comment (jklukas) [reply src/main/scala/com/mozilla/telemetry/sinks/HttpSink.scala]: [Spark's docs on structured streaming](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html) use the terminology "Foreach sink", so I think the name here is in keeping with the spirit of Spark's terminology. "HttpSink" feels to me like a good description of the intent of the class.

That said, working with this PR is my first real exposure to the Spark Streaming API for writing. If you think "HttpSink" could cause confusion for someone deeply familiar with Spark Streaming, we can change it.

**Comment by jklukas** [comment]
Pushed a new commit to add additional Scaladoc to address your points about potentially Amplitude-specific code in HttpSink.

File: src/main/scala/com/mozilla/telemetry/sinks/HttpSink.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (akkomar) [reply src/main/scala/com/mozilla/telemetry/sinks/HttpSink.scala]: Sorry, I wasn't clear enough - I was thinking about this specific set.
I was under impression that they have some non-standard semantics for status codes. If they don't, which their docs seem to confirm, then it should be fine.

> From my understanding, this list of codes is the complete set you'd want to retry on for any API.
I think there are some other retryable ones, e.g. 409.

[the remark this answers is not in the document]
Comment (jklukas) [reply src/main/scala/com/mozilla/telemetry/sinks/HttpSink.scala]: From my reading, the only non-standard part of Amplitude's behavior is the 429 is specific to device IDs. I'm planning to implement their suggestion of partitioning by client ID  to limit the impact of per-device throttling as part of my next PR here.

I will add 409 to this list, though. I agree that sounds like a generally retriable code.

**Pull request merged**